File: app/src/main/java/ru/vyaacheslav/suhov/imeit/MainActivity.java
package ru.vyaacheslav.suhov.imeit;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import java.util.Calendar;
import java.util.Objects;

import ru.vyaacheslav.suhov.imeit.ftagments.CallTime;
import ru.vyaacheslav.suhov.imeit.ftagments.Exzam;
import ru.vyaacheslav.suhov.imeit.ftagments.Session;
import ru.vyaacheslav.suhov.imeit.maps.MapsFragment;

public class MainActivity extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener {

    public ActionBarDrawerToggle toggle;
    public java.util.Calendar calendar;
    public FragmentTransaction FT;
    private FragmentManager FM;
    private SharedPreferences   prefs;
    public String typeTheme,typeWeek, typeGroupe;
    private Intent settingsIntent;
    public Toolbar tb;
    public NavigationView nv;
    private int weekYear = Calendar.getInstance().get(Calendar.WEEK_OF_YEAR);
    private Menu menu;
    private DrawerLayout dl;

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        prefs = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
        typeTheme = prefs.getString(getString(R.string.pref_theme), "");
        typeWeek = prefs.getString(getString(R.string.week_i), "");
        typeGroupe = prefs.getString(getString(R.string.pref_groupe), "");

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            if (Objects.equals(typeTheme, "Светлая")) {
                setTheme(R.style.ThemeWrithe);
            }
            else {
                setTheme(R.style.ThemeDark);
            }
        }

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Интент для запуска окна с настройками
        SharedPreferences prefs = getPreferences(MODE_PRIVATE);
        if (prefs.getBoolean("isFirstRun", true)) {
            Intent intent = new Intent(MainActivity.this, SettingsPref.class);
            startActivity(intent);
        }
        prefs.edit().putBoolean("isFirstRun", false).apply();

        // Стандартная инициализация компонетов
        tb = findViewById(R.id.toolbar);
        nv = findViewById(R.id.shitstuff);
        dl = findViewById(R.id.drawerLayout);

        calendar = Calendar.getInstance();
        setSupportActionBar(tb);
        //ActionBarToggle - Это иконка DrawerLayout
        toggle = new ActionBarDrawerToggle(this, dl, tb, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        dl.addDrawerListener(toggle);
        toggle.syncState();

        // Инициализация транзакции фрагментов
        FM = getSupportFragmentManager();
        FT = FM.beginTransaction();
        FT.replace(R.id.containerView, new TabFragment()).commit(); // По умолчанию - расписание.

        // Обработка нажатий на пункты меню.
        nv.setNavigationItemSelectedListener(this);

        settingsIntent = new Intent(MainActivity.this, SettingsPref.class);
        loadName(); // Загрузка имени группы согласно настройкам;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        this.menu = menu;
        getMenuInflater().inflate(R.menu.menu, menu);
        WeekIndicator();
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.settings:
                startActivity(settingsIntent);
                finish();
                return true;
            case R.id.weekIndicator:
                WeekIndicator();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    public void WeekIndicator() {
        if(typeWeek.equals("Числитель")){

            if(weekYear% 2 == 0){
                Toast.makeText(getApplicationContext(), "Текущая неделя: Числитель", Toast.LENGTH_SHORT).show();
                menu.getItem(0).setIcon(getResources().getDrawable(R.drawable.ch));
            }
            else {
                Toast.makeText(getApplicationContext(), "Текущая неделя: Знаменатель", Toast.LENGTH_SHORT).show();
                menu.getItem(0).setIcon(getResources().getDrawable(R.drawable.ze));
            }
        }
        else {
            if(weekYear% 2 == 0){
                Toast.makeText(getApplicationContext(), "Текущая неделя: Знаменатель", Toast.LENGTH_SHORT).show();
                menu.getItem(0).setIcon(getResources().getDrawable(R.drawable.ze));
            }
            else {
                Toast.makeText(getApplicationContext(), "Текущая неделя: Числитель", Toast.LENGTH_SHORT).show();
                menu.getItem(0).setIcon(getResources().getDrawable(R.drawable.ch));
            }
        }
    }

    public void loadName() {
        // В качестве подзоголовка берем имя выбранной группы.
        typeGroupe = prefs.getString(getString(R.string.pref_groupe), "");
        MainActivity.this.tb.setSubtitle(typeGroupe);
        // Если пользователь очистил память в окне настроек, то строка будет пуста. Делаем проверку.
        if ((typeGroupe.length() == 0)){
            startActivity(settingsIntent);
            finish();
        }
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.main_tab:
                FragmentTransaction ft1 = FM.beginTransaction();
                ft1.replace(R.id.containerView, new TabFragment()).commit();
                loadName();
                break;
            case R.id.map:
                MainActivity.this.tb.setSubtitle("Учебные корпуса");
                FragmentTransaction ft2 = FM.beginTransaction();
                ft2.replace(R.id.containerView, new MapsFragment()).commit();
                break;
            case R.id.time_alarm:
                MainActivity.this.tb.setSubtitle("Время звонков");
                FragmentTransaction ft3 = FM.beginTransaction();
                ft3.replace(R.id.containerView, new CallTime()).commit();
                break;
            case R.id.exzam:
                MainActivity.this.tb.setSubtitle("Экзаменационная сессия");
                FragmentTransaction ft4 = FM.beginTransaction();
                ft4.replace(R.id.containerView, new Exzam()).commit();
                break;
            default:
                break;
        }
        dl.closeDrawer(GravityCompat.START);
        return true;

    }
}

File: app/src/main/java/ru/vyaacheslav/suhov/imeit/MainActivity.kt
package ru.vyaacheslav.suhov.imeit

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import ru.vyaacheslav.suhov.imeit.util.gone
import ru.vyaacheslav.suhov.imeit.util.visible
import ru.vyaacheslav.suhov.imeit.view.ftagments.auth.FragmentLogin
import ru.vyaacheslav.suhov.imeit.view.ftagments.calls.CallSetupFragment
import ru.vyaacheslav.suhov.imeit.view.ftagments.other.BottomNavFragment
import ru.vyaacheslav.suhov.imeit.view.ftagments.other.FragmentEmptyGroup
import ru.vyaacheslav.suhov.imeit.view.ftagments.schedule.SchedulePagerFragment
import ru.vyaacheslav.suhov.imeit.view.view.UpToolbar
import ru.vyaacheslav.suhov.imeit.viewmodel.MainViewModel
import kotlinx.android.synthetic.main.activity_main.*
import ru.vyaacheslav.suhov.imeit.util.ErrorEvent.ERROR_CONNECTIONS
import ru.vyaacheslav.suhov.imeit.util.ErrorEvent.ERROR_CREATE_ACCOUNT
import ru.vyaacheslav.suhov.imeit.util.ErrorEvent.ERROR_EMAIL
import ru.vyaacheslav.suhov.imeit.util.ErrorEvent.ERROR_LOGIN
import ru.vyaacheslav.suhov.imeit.util.ErrorEvent.ERROR_PASS
import ru.vyaacheslav.suhov.imeit.util.ErrorEvent.ERROR_SERVER
import ru.vyaacheslav.suhov.imeit.util.toast

class MainActivity : AppCompatActivity() {

    private lateinit var model: MainViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(bottom_bar)

        model = ViewModelProviders.of(this@MainActivity)[MainViewModel::class.java]

        model.observeFirstRun(this@MainActivity, Observer {
            if (it) {
                model.setFirstRun(false)
                loadStartFragment()
            }
        })

        model.observeAuth(this@MainActivity, Observer {
            if (!it) FragmentLogin().show()
            visibleUI(it)
        })


        if (model.isAuth) loadStartFragment() else FragmentLogin().show()
        observeErrorMsg()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu, menu); return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                val navDrawerFragment = BottomNavFragment()
                navDrawerFragment.show(supportFragmentManager, navDrawerFragment.tag)
            }
            R.id.group_setup -> showDialogSelectGroup()
            R.id.bells_pref -> CallSetupFragment().show()
        }
        return true
    }

    private fun loadStartFragment() {
        if (model.isExistsGroup()) SchedulePagerFragment().show() else FragmentEmptyGroup().show()
    }

    private fun showDialogSelectGroup() {
        AlertDialog.Builder(this@MainActivity)
                .setTitle(R.string.group)
                .setNeutralButton(resources.getString(R.string.cancel)) { d, _ -> d.cancel() }
                .setSingleChoiceItems(model.getListGroups(), model.getSelectedId())
                { d, i ->
                    model.setSelectedId(i)
                    loadStartFragment()
                    d.cancel()
                }
                .create()
                .show()
    }

    private fun showDialogExit() {
        AlertDialog.Builder(this@MainActivity)
                .setTitle(resources.getString(R.string.exit_of_app))
                .setNeutralButton(resources.getString(R.string.cancel)) { d, _ -> d.cancel() }
                .setPositiveButton(resources.getString(R.string.yes)) { d, _ -> d.dismiss(); finish() }
                .create()
                .show()
    }

    private fun Fragment.show() {
        supportFragmentManager
                .beginTransaction()
                .replace(R.id.container, this@show, this@show.tag)
                .commit()
    }

    private fun visibleUI(v: Boolean) {

        val toolbar: UpToolbar = findViewById(R.id.toolbar)
        val bottomView: CoordinatorLayout = findViewById(R.id.bottom_view)

        arrayOf(toolbar, bottomView).forEach { if (v) it.visible() else it.gone() }
    }

    // Вынести в отдельное вью с ошибками!! ^D-D^

    private fun observeErrorMsg() {

        model.observeErrorMsg(this@MainActivity, Observer {
            val msg: Int = when (it) {
                ERROR_CONNECTIONS -> R.string.error_connections
                ERROR_PASS -> R.string.error_pass
                ERROR_EMAIL -> R.string.error_email
                ERROR_LOGIN -> R.string.error_login
                ERROR_SERVER -> R.string.error_server
                ERROR_CREATE_ACCOUNT -> R.string.error_create_account
                else -> 0
            }
            toast(this, msg)
        })
    }

    override fun onBackPressed() {
        showDialogExit()
    }
}